### PR TITLE
Use AutoSave in daily docket

### PR DESCRIPTION
**Pull request by slj**
Connects #2475

### CODE REVIEW

File: client/app/hearings/DailyDocketContainer.jsx [hunk 1]
```diff
@@ -40,26 +41,27 @@ export class DailyDocketContainer extends React.Component {
...
-        <div className="cf-sg-loader">
-          <LoadingContainer color={AppConstants.LOADING_INDICATOR_COLOR_HEARINGS}>
-            <div className="cf-image-loader">
-            </div>
-            <p className="cf-txt-c">Loading dockets, please wait...</p>
-          </LoadingContainer>
-        </div>
+      return <div className="loading-dockets">
+        <div>{loadingSymbolHtml('', '50%', LOADING_INDICATOR_COLOR_HEARING_PREP)}</div>
+        <div>Loading hearings, please wait...</div>
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: Can you sync with @nolawi on these changes? I believe he just fixed the positioning on this recently.

File: client/app/hearings/actions/Dockets.js [hunk 2]
```diff
@@ -21,59 +21,137 @@ export const handleServerError = (err) => ({
...
 
-export const setTranscriptRequested = (hearingIndex, transcriptRequested, date) => ({
-  type: Constants.SET_TRANSCRIPT_REQUESTED,
-  payload: {
-    hearingIndex,
-    transcriptRequested,
-    date
-  }
-});
+    eval(`
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: why do we need an `eval` here? Eval is incredibly dangerous, and is generally never used. Can we assign the inner function to a variable instead?

File: client/app/hearings/actions/Dockets.js [hunk 3]
```diff
@@ -21,59 +21,137 @@ export const handleServerError = (err) => ({
   }
 });
 
-export const setNotes = (hearingIndex, notes, date) => ({
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: This file is really difficult to review because of all the unnecessary diffs. The functions `setNotes`, `setDisposition` etc should not show up as changed if no change was desired. 

Can you push a new commit that redoes this file so only the true changes show up in the diff?

File: client/app/hearings/actions/Dockets.js [hunk 4]
```diff
@@ -21,59 +21,137 @@ export const handleServerError = (err) => ({
...
+      }, ${milliseconds});`
+    );
+    // filter out edited:true + pass that to saveData()?
+
+  };
+/* eslint-enable no-unused-vars */
+/* eslint-enable no-eval */
+
+/* eslint-disable no-unused-vars */
+export const saveData = (data, dispatch, url = '/hearings/save_data') => {
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: can we remove the `url` param for now, since it's not being used yet

File: client/app/hearings/actions/Dockets.js [hunk 5]
```diff
@@ -21,59 +21,137 @@ export const handleServerError = (err) => ({
...
-export const setNotes = (hearingIndex, notes, date) => ({
-  type: Constants.SET_NOTES,
-  payload: {
-    hearingIndex,
-    notes,
-    date
-  }
-});
+/* eslint-disable no-unused-vars */
+/* eslint-disable no-eval */
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: we should be able to remove both of these eslint disables. Which unused vars are causing the first lint error?

File: client/app/hearings/reducers/index.js [hunk 6]
```diff
@@ -82,6 +82,11 @@ export const hearingsReducers = function(state = mapDataToInitialState(), action
       worksheet: { comments: { $set: action.payload.comments } }
     });
 
+  case Constants.TOGGLE_SAVING:
```
Comment (joofsh) [inline client/app/hearings/reducers/index.js]: nice 👍

File: client/app/hearings/actions/Dockets.js (hunk 2 above)
Comment (slj) [reply client/app/hearings/actions/Dockets.js]: `eval` may not be necessary, but I don't think there's a simpler approach.  It's incredibly dangerous when the string being evaluated is comprised of user input, but that's not the case here.  Can you give me an example of your inner function approach?

File: client/app/hearings/actions/Dockets.js [hunk 7]
```diff
@@ -1,5 +1,43 @@
...
+  (dispatch, getState) => {
+
+    dispatch(action);
+
+    lastChangeTimestamp = Number(new Date());
+
+    // create a function that, after a moment (default 1000ms),
+    // checks if there haven't been any more calls to updateHearing().
+    // If true (meaning the time at which the function runs matches
+    // the time at which it was created), save the new data.
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: can you give me more detail in what you're trying to accomplish here? I think what you might want to use here is [lodash debounce](https://lodash.com/docs/4.17.4#debounce)

File: client/app/hearings/actions/Dockets.js [hunk 8]
```diff
@@ -1,5 +1,43 @@
...
+
+    eval(`
+      setTimeout(() => {
+        if (${lastChangeTimestamp} === lastChangeTimestamp) {
+          dispatch({type:'TOGGLE_SAVING'});
+          saveData(getState().dockets, dispatch);
+        }
+      }, ${milliseconds});`
+    );
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: For more information on eval and why we shouldn't use it see [this doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval), in specific: 

> eval() is a dangerous function, which executes the code it's passed with the privileges of the caller. If you run eval() with a string that could be affected by a malicious party, you may end up running malicious code on the user's machine with the permissions of your webpage / extension. More importantly, third party code can see the scope in which eval() was invoked, which can lead to possible attacks in ways to which the similar Function is not susceptible.
> 
> eval() is also generally slower than the alternatives, since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
> 
> There are safer (and faster!) alternatives to eval() for common use-cases.

File: client/app/hearings/actions/Dockets.js [hunk 9]
```diff
@@ -30,7 +82,7 @@ export const setNotes = (hearingIndex, notes, date) => ({
   }
 });
 
-export const setDisposition = (hearingIndex, disposition, date) => ({
+export const setDisposition = (hearingIndex, disposition, date) => updateHearing({
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: okay, now with the improved diff I understand what you're doing here, but I don't think this is the right approach. We don't want to couple the action creators with dispatching other actions. This is highly coupled & confusing.

We should treat creating & dispatching actions (updating the redux store) separately from how we trigger saves to the server. This is partially why we created the `edited` flag in the redux store. A separate function/loop should be able to run that looks at the redux store and determines what should be sent to the backend

File: client/app/hearings/actions/Dockets.js [hunk 10]
```diff
@@ -1,5 +1,43 @@
...
+    lastChangeTimestamp = Number(new Date());
+
+    // create a function that, after a moment (default 1000ms),
+    // checks if there haven't been any more calls to updateHearing().
+    // If true (meaning the time at which the function runs matches
+    // the time at which it was created), save the new data.
+
+    eval(`
+      setTimeout(() => {
+        if (${lastChangeTimestamp} === lastChangeTimestamp) {
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: are there records where `edited` is true

File: client/app/hearings/actions/Dockets.js [hunk 11]
```diff
@@ -1,5 +1,43 @@
...
+
+    lastChangeTimestamp = Number(new Date());
+
+    // create a function that, after a moment (default 1000ms),
+    // checks if there haven't been any more calls to updateHearing().
+    // If true (meaning the time at which the function runs matches
+    // the time at which it was created), save the new data.
+
+    eval(`
+      setTimeout(() => {
```
Comment (joofsh) [inline client/app/hearings/actions/Dockets.js]: setInterval

File: client/app/hearings/actions/Dockets.js (hunk 8 above)
Comment (joofsh) [reply client/app/hearings/actions/Dockets.js]: example API discussed in person

```
<AutoSave
  spinnerColor="green"
  saveAction="foo"
  calculateIfSaveNecessary=function() {
     ...loop through dockets, see if any edited
       dispatch the action to save
  }
  AutoSaveInterval="30"
/>

Aut

```

File: client/app/hearings/DailyDocketContainer.jsx [hunk 12]
```diff
@@ -33,6 +34,61 @@ export class DailyDocketContainer extends React.Component {
     }
   }
 
+  docket = () => {
+    return this.props.dockets[this.props.date].hearings_hash;
+  }
+
+  saveFunction = (dispatch) => {
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: this function looks great in general. Just 1 minor nit: Is there a reason you separate it out here rather than in the `mapDispatchToProps`? It feels like the `save` and `saveFunction` could be merged

File: client/app/hearings/DailyDocketContainer.jsx [hunk 13]
```diff
@@ -33,6 +34,61 @@ export class DailyDocketContainer extends React.Component {
...
+    // This function is expcted to be passed to the "save" property defined in
+    // AutoSave.mapDispatchToProps().  save() calls this function and passes a
+    // ref to dispatch(), enabling this function to fire actions as needed.
+
+    let hearingsToSaveIndeces = [];
+
+    const hearingsToSave = this.docket().filter((hearing, index) => {
+      if (hearing.edited) {
+        hearingsToSaveIndeces.push(index);
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: minor nit: the way you're doing it here is most efficient (only looping through once with the `filter`), but I'm worried it may be confusing since people wouldn't expect the `filter` to have a side effect of building the `hearingsToSaveIndeces` Do you mind splitting it out `hearingsToSave.map(h => h.id)`

File: client/app/hearings/DailyDocketContainer.jsx (hunk 12 above)
Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: Doing it this way prevents AutoSave from having to care about any details of the save - all the details are kept in the component that uses it - this is a more reusable approach IMO

File: client/app/hearings/DailyDocketContainer.jsx (hunk 13 above)
Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: `hearingsToSave.map(h => h.id)` will not help accomplish the goal, because the SET_EDITED_TO_FALSE action needs the hearings' positions in the `hearings_hash` array, not their ids.

While I do not agree that the `filter` side effect is as confusing as you suspect, I will compile the indeces in a separate loop.

File: client/app/hearings/DailyDocketContainer.jsx (hunk 12 above)
Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: Great point 👍

File: client/app/hearings/DailyDocketContainer.jsx (hunk 13 above)
Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: Oh hmm, I didn't catch that. Will add a comment down below

File: client/app/hearings/reducers/index.js [hunk 14]
```diff
@@ -82,6 +82,22 @@ export const hearingsReducers = function(state = mapDataToInitialState(), action
...
+    return update(state, {
+      isSaving: { $set: !state.isSaving }
+    });
+
+  case Constants.SET_EDITED_FLAG_TO_FALSE:
+    return update(state, {
+      dockets: {
+        [action.payload.date]: {
+          hearings_hash: {
+            [action.payload.index]: { edited: { $set: false } }
```
Comment (joofsh) [inline client/app/hearings/reducers/index.js]: Can you sync with @nolawi on the new docket structure. I believe we are changing this to be key/value pairs using the hearing's database id.  I'd rather not use indexes in general if we can avoid it (they are relatively fragile)

Comment (slj) [reply client/app/hearings/reducers/index.js]: Aha.  Ok, will do.

Comment (slj) [reply client/app/hearings/reducers/index.js]: Just pulled master - `hearings_array` is still a true array, not key/value pairs.

**Comment by slj** [comment]
I spoke with @nolawi, who informed me that there's not yet a ticket to change `hearings_array` to resemble the structure proposed in https://github.com/department-of-veterans-affairs/caseflow/issues/3137.  So unless there's anything else outstanding, I think this PR is ready to leave WIP status.

**Comment by nolawi** [comment]
ah yes, the `dockets.json` needs to be updated first _ we are still in placeholder in hearings

File: client/app/hearings/DailyDocketContainer.jsx [hunk 15]
```diff
@@ -55,11 +116,19 @@ export class DailyDocketContainer extends React.Component {
...
 
-    return <DailyDocket
-      veteran_law_judge={this.props.veteran_law_judge}
-      date={this.props.date}
-      docket={this.props.dockets[this.props.date].hearings_array}
-    />;
+    // temporarily use 5 second interval for quicker validation
+    return <div className="cf-hearings-daily-docket-container">
+      <AutoSave
+        intervalInMs={5000}
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: can we please remove this. The AC that needs to be validated specifically says 30 seconds

Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: It also calls it a "configurable property".  In previous conversations I thought we agreed 30 seconds would be a default, so I see no AC violation here.  Nevertheless, I will remove it.

File: client/app/components/AutoSave.jsx [hunk 16]
```diff
@@ -17,9 +17,13 @@ export class AutoSave extends React.Component {
...
     if (!window.onbeforeunload) {
       window.onbeforeunload = () => {
-        this.props.doBeforeWindowCloses();
+        this.props.save(this.props.saveFunction);
       };
     }
+
+    setInterval(() => {
+      this.props.save(this.props.saveFunction);
+    }, this.props.intervalInMs || 30000);
```
Comment (joofsh) [inline client/app/components/AutoSave.jsx]: rather than directly code the default here, let's use React's built in [defaultProps](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/components/StatusMessage.jsx#L58-L60)

File: client/app/components/AutoSave.jsx [hunk 17]
```diff
@@ -42,6 +46,9 @@ const mapStateToProps = (state) => ({
 const mapDispatchToProps = (dispatch) => ({
   doBeforeWindowCloses: () => {
     dispatch(this.props.beforeWindowClosesActionCreator());
+  },
+  save: (saveFunction) => {
```
Comment (joofsh) [inline client/app/components/AutoSave.jsx]: I like this concept, but can we move this to the `DailyDocketsContainer`

File: client/test/karma/components/AutoSave-test.js [hunk 18]
```diff
@@ -36,12 +36,12 @@ describe('AutoSave', () => {
       const wrapper = mount(
         <AutoSave
```
Comment (joofsh) [inline client/test/karma/components/AutoSave-test.js]: Can we add tests for the 30 sec save trigger?  And is there a reason the test on line 50 is skipped? If we can't use `asyncTest` that's okay for now, we should still test the component with `setTimeout`

File: client/app/components/AutoSave.jsx [hunk 19]
```diff
@@ -12,14 +12,18 @@ const now = () => {
     replace(/(p|a)m/, '$1.m.');
 };
 
+const autoSaveInterval = { id: 0 };
```
Comment (joofsh) [inline client/app/components/AutoSave.jsx]: for values we want to save, but that should exist outside the react lifestyle, we can store them on the component's `this`. Here's an example from Reader https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/reader/Pdf.jsx#L101

File: client/app/components/AutoSave.jsx [hunk 20]
```diff
@@ -12,14 +12,18 @@ const now = () => {
...
 
   componentDidMount = () => {
     if (!window.onbeforeunload) {
       window.onbeforeunload = () => {
-        this.props.doBeforeWindowCloses();
+        this.props.save();
       };
     }
+
+    autoSaveInterval.id = setInterval(() => this.props.save(), this.props.intervalInMs);
```
Comment (joofsh) [inline client/app/components/AutoSave.jsx]: Do we expect to have multiple things stored in the object here? Otherwise I think it would be simpler as just a single value `autoSaveInterval = setInterval(() => this.props.save(), this.props.intervalInMs);`

File: client/app/hearings/DailyDocketContainer.jsx [hunk 21]
```diff
@@ -32,6 +34,58 @@ export class DailyDocketContainer extends React.Component {
     }
   }
 
+  docket = () => {
+    return this.props.dockets[this.props.date].hearings_array;
+  }
+
+  save = () => {
+    const dispatch = this.props.getDispatch();
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: this feels awkward. Can we instead move the whole `save` function down to the `mapStateToDispatch` on line 135?

File: client/test/karma/components/AutoSave-test.js [hunk 22]
```diff
@@ -36,24 +36,53 @@ describe('AutoSave', () => {
...
     });
   });
 
-  xit('calls an action creator before window closes', asyncTest(async () => {
-    mount(
-      <AutoSave beforeWindowClosesActionCreator={actionCreator} />
-    );
-    window.close();
-    await pause();
+  context('calls save', () => {
```
Comment (joofsh) [inline client/test/karma/components/AutoSave-test.js]: wow! Excellent tests 😍 Great use of `sinon.useFakeTimers()` 👍

File: client/app/hearings/DailyDocketContainer.jsx (hunk 21 above)
Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: Can you go into more detail. What do you mean by "will not work"? What have you tried so far that hasn't worked?

File: client/app/components/AutoSave.jsx [hunk 23]
```diff
@@ -12,18 +12,21 @@ const now = () => {
     replace(/(p|a)m/, '$1.m.');
 };
 
-const autoSaveInterval = { id: 0 };
-
 export class AutoSave extends React.Component {
 
+  constructor(props) {
+    super(props);
+    this.setIntervalId = 0;
```
Comment (joofsh) [inline client/app/components/AutoSave.jsx]: let's set this to null, `0` is a real value where we really want no value in this case. You in theory could not set/define it at all in the constructor, but I like the clarity added by defining it (so we know it's something used later in the component's lifecycle)

File: client/app/hearings/DailyDocketContainer.jsx (hunk 21 above)
Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: As is, `save()` will not work if moved to `mapDispatchToProps` near line 135 because it relies on values in `this.props`, which is not accessible inside `mapDispatchToProps`.  Furthermore, `ownProps` does not contain the values used in `save()`.  I'm happy to discuss this in person going forward.

Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: okay, I dug into the code a little bit and I understand what you mean now. At a high level, we should not be passing around `dispatch` like we do. Ideally within the `DailyDocketContainer` it would have no awareness of or reference to `dispatch`.  The more common pattern is to isolate all dispatch-related logic to the `mapStateToProps` and `mapDispatchToProps` functions. That way there is a clean separation of concerns. The component itself simply called functions that it has been passed in, and is not responsible for their inner workings. We can accomplish this with a few refactors:

- Remove `dockets` as a property into the component. We should replace this with `docket`. The DailyDocketComponent only cares about its one specific `docket`, not the collection of all `dockets`
- Always call `getDockets` when the component mounts. We should move the conditional check to `mapDispatchToProps`. When the function is called (on component mount) we can check if the dockets array already exists in the redux store. If it does, the function call can no-op.

With those refactors, we've "thinned" the need to have the `save` function exist within the component. There are a couple ways to do this, but I would recommend:

1) Move the `save` function to `mapDispatchToProps`
2) Use Redux's `mergeProps` 3rd argument to connect. See the [api docs](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options) for more information. Here is an example of calling the function with a specific prop:

```
function mergeProps(stateProps, dispatchProps, ownProps) {
  return Object.assign({}, ownProps, stateProps, dispatchProps, {
    addTodo: (text) => dispatchProps.addTodo(ownProps.userId, text)
  })
}

export default connect(mapStateToProps, actionCreators, mergeProps)(TodoApp)
```
So basically we can add `save: () => dispatchProps.save(hearings)` in the mergeProps. This overrides the existing version from `dispatchProps` with our

Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: Looking through this component, there are some other refactors that might be helpful in avoiding this sort of confusion in the future:

- Remove `dockets` as a property into the component. We should replace this with `docket`. The DailyDocketComponent only cares about its one specific `docket`, not the collection of all `dockets`
- Always call `getDockets` when the component mounts. We should move the conditional check to `mapDispatchToProps`. When the function is called (on component mount) we can check if the dockets array already exists in the redux store. If it does, the function call can no-op.

Do you mind making a tech-improvement ticket to make these changes? This should be the next thing you work on after this is merged

File: client/app/hearings/DailyDocketContainer.jsx [hunk 24]
```diff
@@ -81,6 +137,6 @@ export default connect(
 
 DailyDocketContainer.propTypes = {
   veteran_law_judge: PropTypes.object.isRequired,
-  dockets: PropTypes.object,
```
Comment (joofsh) [inline client/app/hearings/DailyDocketContainer.jsx]: this is still a property of the component. Can we re-add this please?

Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: This property is not passed into the component. See https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/hearings/index.jsx#L61-63

Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: Ah, but `dockets` is via `mapStateToProps`. It's still a property that is used internally within the component, and so should be documented here

Comment (joofsh) [reply client/app/hearings/DailyDocketContainer.jsx]: In fact, `serverError` should also be added here

Comment (slj) [reply client/app/hearings/DailyDocketContainer.jsx]: It was already there in my branch

**Comment by slj** [comment]
Thx man
